### PR TITLE
Add ability to zoomout from base state

### PR DESCRIPTION
**Pull request by jessyec-s**
This patch makes it possible to zoom out from the chart's base zoom state of 100% by decreasing the lane height of each thread. 

According to the number of threads in the chart, the amount the chart can zoom out from its base state differs.  For example decreasing lane height is not possible if the total space all lanes take up is less than the total visible height.  In this case lanes adapt their heights to fill the entire screen, so "zooming out" from this state would create empty space, and therefore is not possible.

Because of this variability,  I have made it so the chart can zoom out from the its base state only if the visible lanes take up less than or equal to 50% of the total chart canvas.  This number was just chosen arbitrarily through trial and error checks, so it can be changed if needed.

Currently, if the chart cannot zoom out anymore because doing such would result in a too small lane height, the chart holds its state as the scale is clicked and decreased to 0%.

A solution to make the scale more accurate would be to make the scale reflect the amount the chart can actually zoom out. This can be done by either making the size of the zoom out increments below 100% variable and depend on how many zoom out steps the chart can do.  Or the 100% base state of the scale can be visually lowered or raised on the scale, depending on how many zoom out steps the chart can take at a fixed increment of 1. 

However, I'm not sure if this variability would look confusing if the scale keeps on changing if different amounts of threads keep on being selected.

Also, I cant decide if this zoom out from base state functionality is clear in general, as it makes zooming in higher than 100% zoom the x axis and zooming out below 100% zoom the y axis of the chart.

Let me know what you think.

### CODE REVIEW

File: application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ThreadsPage.java [hunk 1]
```diff
@@ -318,6 +319,7 @@ protected IXDataRenderer getChartRenderer(IItemCollection itemsInTable, Histogra
 							.collect(Collectors.toList());
 					chartCanvas.setNumItems(this.threadRows.size());
 					textCanvas.setNumItems(this.threadRows.size());
+					DisplayToolkit.inDisplayThread().execute(() -> displayBar.adjustScaleBounds());;
```
Comment (aptmac) [inline application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ThreadsPage.java]: extra `;`

File: application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/misc/ChartTextCanvas.java [hunk 2]
```diff
@@ -67,7 +67,9 @@
 import org.openjdk.jmc.ui.misc.PatternFly.Palette;
 
 public class ChartTextCanvas extends Canvas {
-	private static int MIN_LANE_HEIGHT = 50;
+	private static final int DEFAULT_LANE_HEIGHT = 50;
```
Comment (aptmac) [inline application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/misc/ChartTextCanvas.java:70]: Not sure how I feel about these two variables having the same name .. maybe the `final` one could be the `DEFAULT` and the practical one be just `LANE_HEIGHT`? Or something similar to that?

**Review by aptmac — COMMENTED** [review]
This is really cool! It's nice to be able to see more threads on the same view, and it looks to be working well.

Having tried it, I'm wondering if this functionality should be independent of the traditional "zoom". This would allow for this "zoomed out" chart to use the existing zoom controls, but also from a usability standpoint it takes many clicks to go from 0% -> 100% where the zooming actually happens, if the adjusted lane heights became the new "100%" then there wouldn't be as many actions required from fully zoomed out -> initial zoom step. Also from a customization point I think it's pretty neat.

Having said that I think Muru is out of the office for a bit, I sent an e-mail asking about adding the date to the time filter and got an auto-reply. I'm not completely sure what this should look like .. my initial thought would be a small widget anchored in the bottom left that would simply have two buttons "+" and "-" for increasing and decreasing the lane height respectively.

What are your thoughts on that?

File: application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/misc/ChartTextCanvas.java (hunk 2 above)
Comment (jessyec-s) [reply application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/misc/ChartTextCanvas.java:70]: Hmm what two variables are you talking about? I set the laneHeight variable to be initialised to the DEFAULT_LANE_HEIGHT value - is this what you are referring to ?

**Comment by jessyec-s** [comment]
> This is really cool! It's nice to be able to see more threads on the same view, and it looks to be working well.
> 
> Having tried it, I'm wondering if this functionality should be independent of the traditional "zoom". This would allow for this "zoomed out" chart to use the existing zoom controls, but also from a usability standpoint it takes many clicks to go from 0% -> 100% where the zooming actually happens, if the adjusted lane heights became the new "100%" then there wouldn't be as many actions required from fully zoomed out -> initial zoom step. Also from a customization point I think it's pretty neat.
> 
I like this idea I think it will make the implementation a lot more simple 
> Having said that I think Muru is out of the office for a bit, I sent an e-mail asking about adding the date to the time filter and got an auto-reply. I'm not completely sure what this should look like .. my initial thought would be a small widget anchored in the bottom left that would simply have two buttons "+" and "-" for increasing and decreasing the lane height respectively.
> 
This idea sounds good to me I'll try adding it.

> What are your thoughts on that?

File: application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/misc/ChartTextCanvas.java (hunk 2 above)
Comment (aptmac) [reply application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/misc/ChartTextCanvas.java:70]: Hm disregard that, I think my browser (or eyes) were playing tricks on me when I read through it the first time.

**Comment by jessyec-s** [comment]
Ok, in https://github.com/aptmac/jmc/commit/d82b7c821bf5be9859c057154fc8c2751055cd65, I have added lane height increase and decrease buttons to the bottom left corner of the chart.  The button images are just placeholder as they are a bit blurry right now. 

Maybe the respective button(s) should be disabled if you cannot increase or decrease the lane height for the chart?

**Comment by aptmac** [comment]
Nice! This works really well, the only thing I can think of that should be included would be a brief tooltip description of what the controls do.

**Comment by jessyec-s** [comment]
Added tooltips in https://github.com/aptmac/jmc/pull/51/commits/c63651ab74337c1dc50399aac907342528a42e95

**Review by aptmac — APPROVED** [review]

**Pull request merged**